pipeline/processor/postprocessor.py

```diff
@@ -0,0 +1,663 @@
+        else:
+            custom_attributes = {}
+
+        text_block = {
+            "text": merged_block_text,
+            "type": "merged_text_block",
+            "coords": merged_coords,
+            "path": text_blocks[0]["path"],
+            "text_block_id": text_blocks[0]["text_block_id"] + "_merged",
+            "custom_attributes": custom_attributes,
+        }
+        return text_block
+
+    def process(self, document: Document) -> Document:
+        """
+        Iterate through a document and merge text blocks that have been separated due to styling elements.
+
+        Args:
+            document:
+
+        Returns:
+                A new dict object with styling info added.
+        """
+        new_document = deepcopy(document)
+        new_document = new_document.to_dict()
+
+        for page in new_document["pages"]:
+            # If page blocks do not have a path (because they're from the embedded text extractor), skip them.
+            # TODO: This is a hack. We should be able to handle this better.
+            if page["text_blocks"][0]["path"] is None:
+                continue
+            # Count repeated paths since blocks with custom styling (subscript, superscript, underline)
+            # have separate elements in the same text block.
+            path_counts = Counter(
+                [tuple(block["path"]) for block in page["text_blocks"]]
+            )
+
+            # TODO: This logic does not always work. For instance, cclw-9460 separates
+            duplicated_paths = [
+                path for path, count in path_counts.items() if count > 1
+            ]
+            try:
+                for path in duplicated_paths:
+                    text_block_idxs, text_blocks_to_merge = list(
+                        zip(
+                            *[
+                                (idx, block)
+                                for idx, block in enumerate(page["text_blocks"])
+                                if tuple(block["path"]) == path
+                            ]
+                        )
+                    )
+                    merged_text_block = self.merge_text_blocks(text_blocks_to_merge)
+                    page["text_blocks"] = (
+                        page["text_blocks"][0 : text_block_idxs[0]]
+                        + [merged_text_block]
+                        + page["text_blocks"][text_block_idxs[-1] + 1 :]
+                    )
+            except TypeError:
+                pass
+
+        new_document = Document.from_dict(new_document)
+        return new_document
+
+
+class AdobeListGroupingPostProcessor(PostProcessor):
+    """Further processing of processed outputs from the Adobe API (handling cases not easily
+    handled by the API itself)."""
+
+    regex_pattern = r"L\[?\d?\]?$"
+
+    def __init__(self):
+        super(AdobeListGroupingPostProcessor, self).__init__()
+
+    @staticmethod
+    def _find_first_occurrence(regex_pattern: str, list_of_strings):
```

Comment:
As this method and the method below are always used to find the *list* occurrences, and are almost identical in their implementation, I'd suggest the following changes:
• rename `self.regex_pattern` to `self.list_regex_pattern`
• `_find_all_occurrences` -> `_find_all_list_occurrences`
• delete `_find_first_occurrence` and instead use `self._find_all_list_occurrences(list_of_strings)[0]` in your code. Alternatively rename `_find_first_list_occurrence` and set its implementation to use `_find_all_list_occurrences`